### PR TITLE
Port Segmentation Camera material to Hlms pieces

### PR DESCRIPTION
**Pull request by darksylinc**
This PR _partially_ implements #504 but is ready for inclusion.

I could add more stuff but I believe it's better to split it into smaller PRs to make review and testing more manageable

# 🦟 Bug fix

See #504 
(it doesn't close it, but merging this PR gets us closer)

## Summary

What was done:

- Ogre2SegmentationMaterialSwitcher now longer switches materials
  - It's still an important class because it sets the right values to each Renderable
  - But now the naming is obviously misleading
  - Hlms customizations take care of the shader part; which allows:
    - Faster performance for not having to switch materials twice per frame (Ogre-Next is not optimized for fast shader switching)
    - Features like skeletal skinning would not work with the old material switcher
 - Harmonized shader customizations. `HlmsPbsTerraShadows` and `Ogre2IgnHlmsCustomizations` were conflicting. They can now be applied simultaneously.
 - `HlmsPbs` and `HlmsUnlit` are now derived by `IgnHlmsPbs` & `IgnHlmsUnlit` to manage all listener-based customizations in a centralized way and provide per-renderable customizations (listener-based customizations can only affect objects in a global state; but cannot send data per renderable which is what Segmentation Camera needs)
 - `Ogre2IgnHlmsCustomizations` had PUBLIC visibility instead of PRIVATE. This was fixed. `Ogre2RenderEngine::HlmsCustomizations()` can return a `Ogre2IgnHlmsCustomizations&` reference but its header was never exposed to users
 - `Ogre2RenderEnginePrivate` is now explicitly HIDDEN (otherwise we get compiler warnings)

What is missing and will be implemented in follow up PRs:

 - Rename `Ogre2IgnHlmsCustomizations` to `Ogre2SphericalClipMinDistance`
 - Perform the same tasks for other material switchers such as `Ogre2SelectionBuffer`, `Ogre2ThermalCameraMaterialSwitcher`, `Ogre2LaserRetroMaterialSwitcher` and possibly others
 - Apply the same customizations to `HlmsTerra` (i.e. heightmaps)
   - This is the main endgoal, given that heightmaps can't be easily implemented through material switchers

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by darksylinc** [comment]
# Question:

Should `IgnHlmsPbs` & `IgnHlmsUnlit` be called `Ogre2IgnHlmsPbs` & `Ogre2IgnHlmsUnlit` instead?

File: ogre2/src/Ogre2IgnHlmsPbsPrivate.cc
Comment (ahcorde) [inline ogre2/src/Ogre2IgnHlmsPbsPrivate.cc:20]: alphabetize

File: ogre2/src/Ogre2IgnHlmsPbsPrivate.hh
Comment (ahcorde) [inline ogre2/src/Ogre2IgnHlmsPbsPrivate.hh:102]: return ?

File: ogre2/src/Ogre2IgnHlmsPbsPrivate.hh
Comment (ahcorde) [inline ogre2/src/Ogre2IgnHlmsPbsPrivate.hh:111]:
```suggestion
    /// \param _shadowNode
    /// \param _casterPass see base class
    /// \param _dualParaboloid
```

Comment (ahcorde) [reply ogre2/src/Ogre2IgnHlmsPbsPrivate.hh:111]: Use in `\param`: `[in]`, `[inout]` or `[out]`

File: ogre2/src/Ogre2IgnHlmsUnlitPrivate.hh
Comment (ahcorde) [inline ogre2/src/Ogre2IgnHlmsUnlitPrivate.hh:98]: `\return` ?

File: ogre2/src/Ogre2IgnHlmsUnlitPrivate.hh
Comment (ahcorde) [inline ogre2/src/Ogre2IgnHlmsUnlitPrivate.hh:108]: _shadowNode and _dualParaboloid

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by iche033 — COMMENTED** [review]
Nice, it's working well for me. 

One comment / question about rendering solid color mode - this is more important for your follow-up PR when making this work for Ogre2MaterialSwitcher (used by selection buffer). There was logic in the original implementation to check for overlay material as it's needed for the gizmo visual, i.e. the RGB axes need to be rendered as overlay so that it can be clicked on by the mouse. Would the new setup work for this use case?

**Comment by darksylinc** [comment]
> One comment / question about rendering solid color mode - this is more important for your follow-up PR when making this work for Ogre2MaterialSwitcher (used by selection buffer).

If you mean this bit of code:

At init:
```cpp
// plain overlay material
this->plainOverlayMaterial =
    this->plainMaterial->clone("plain_color_overlay");
if (!this->plainOverlayMaterial->getTechnique(0) ||
    !this->plainOverlayMaterial->getTechnique(0)->getPass(0))
{
  ignerr << "Problem creating selection buffer overlay material"
      << std::endl;
  return;
}
Ogre::Pass *overlayPass =
    this->plainOverlayMaterial->getTechnique(0)->getPass(0);
Ogre::HlmsMacroblock macroblock(*overlayPass->getMacroblock());
macroblock.mDepthCheck = false;
macroblock.mDepthWrite = false;
overlayPass->setMacroblock(macroblock);
```

At evaluation per frame:

```cpp
// check if it's an overlay material by assuming the
// depth check and depth write properties are off.
if (!datablock->getMacroblock()->mDepthWrite &&
    !datablock->getMacroblock()->mDepthCheck)
{
  subItem->setMaterial(this->plainOverlayMaterial);
}
```

It seems like the actual basically boils down to:

```cpp
if( hasTransparency( originalMaterial ) ) // Basically checks mDepthWrite = false & mDepthCheck = false
{
   useSameButTransparentMaterial(); // Basically sets mDepthWrite = false & mDepthCheck = false
}
```

If that's the case then it should work, because the Hlms customization will honour the original material's settings.

**However this does raise an important thing to point out:** This PR will not always produce the exact same output.

The reason is that the original would take a Renderable, apply either `plainMaterial` or `plainOverlayMaterial` and render the frame with it.

However this customization will honour the (possibly complex, possibly not) macro- and blendblock settings from the original material. In some cases this is desirable and the previous implementation would actually be considered buggy.

In other cases it may not be desirable. Probably alpha/additive blending is not desirable. A possible solution would be to clone the materials of those objects (and only those) that are trying to apply blending (with one that does not).

**Comment by iche033** [comment]
yes that's the use case I'm referring to.

> However this customization will honour the (possibly complex, possibly not) macro- and blendblock settings from the original material. In some cases this is desirable and the previous implementation would actually be considered buggy.

hmm going back to the selection buffer use case, if we have a renderable with a semi-transparent material (or material with some alpha blendblock), would its color change if it's in front of another solid object? For selection buffer / mouse picking to work, we'll need to make sure it's rendered the same unique solid color that's assigned to it.

File: ogre2/src/Ogre2IgnHlmsPbsPrivate.cc
Comment (darksylinc) [reply ogre2/src/Ogre2IgnHlmsPbsPrivate.cc:20]: I fixed the other errors, but I'm not sure I understand what I need to do here?

The whole snippet is:

```cpp
#include "Ogre2IgnHlmsPbsPrivate.hh"

#include "Terra/Hlms/PbsListener/OgreHlmsPbsTerraShadows.h"

#include <ignition/common/Filesystem.hh>
#include <ignition/common/Util.hh>
```

Alphabetically:

```cpp
#include "Ogre2IgnHlmsPbsPrivate.hh"

#include <ignition/common/Filesystem.hh>
#include <ignition/common/Util.hh>
#include "Terra/Hlms/PbsListener/OgreHlmsPbsTerraShadows.h"
```

Is that what you're asking?

Cheers

**Comment by darksylinc** [comment]
> hmm going back to the selection buffer use case, if we have a renderable with a semi-transparent material (or material with some alpha blendblock), would its color change if it's in front of another solid object? For selection buffer / mouse picking to work, we'll need to make sure it's rendered the same unique solid color that's assigned to it.

Yeah that was I referring to by "undesirable". Definitely we'll need to turn blending off.

**Review by iche033 — APPROVED** [review]
looks good to me!

**Comment by darksylinc** [comment]
Before merging this, I want to include the fix to turn blending off. That's not much work and should definitely be included in this PR.

**Comment by darksylinc** [comment]
OK I fixed the problem with the blendblocks.

It should be ready for merge now.

**Comment by iche033** [comment]
latest changes look good to me.

**Comment by darksylinc** [comment]
I'm not certain what 'alphabetize' means in this context (https://github.com/ignitionrobotics/ign-rendering/pull/517#discussion_r772178166), thus I don't know what/how to fix

File: ogre2/src/Ogre2IgnHlmsPbsPrivate.cc
Comment (iche033) [reply ogre2/src/Ogre2IgnHlmsPbsPrivate.cc:20]: hmm I think the order should be ok based on [google's style guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) which we've been somewhat loosely following. The related headers are included first in alphabetical order then other libraries headers in alphabetical order. 

Pinging @ahcorde again to see what he has in mind.

**Comment by darksylinc** [comment]
Btw can I ask to merge 6 -> main before merging this PR?

I can rebase the changes.

It's because these Hlms changes is tipping its toes with PRs #522 #523 & #524 which have already been merged but are not yet in main.

**Comment by iche033** [comment]
> Btw can I ask to merge 6 -> main before merging this PR? (preferably ASAP)

created PR #532, will merge when CI turns green

**Comment by iche033** [comment]
Going to merge this to unblock other follow-up work. We can address the remaining style issue later (if that's still considered an issue)

**Pull request merged**
